### ISSUE TITLE
Remove the need to keep implementing life cycle classes on subclasses

### ISSUE THREAD

**Issue by frankpepermans** [dart-lang/angular2#19]
Currently you can extend existing components or directives, but in order for the life cycle hooks to trigger, the subclass must always copy the 'implements' list.

i.e.

```
@Component()
class Foo implements OnDestroy {
    void ngOnDestroy() => 'do something';
}
```

```
@Component()
class Bar extends Foo implements OnDestroy {
    // ngOnDestroy is called
}
```

```
@Component()
class Baz extends Foo {
    // ngOnDestroy is not called
}
```

Bar need to declare that it imlpements OnDestroy, or the ngOnDestroy method is not called.

I've always assumed that this was a downside from the compile-from-TS solution, is there any chance this can be adjusted? (ending up with the Baz class instead of the Bar class)

**Comment by matanlurey** [comment]
So you definitely want the interface in Dart because you'll get static checks if you forget to implement the ngOnDestroy method and you can make a clean (component as OnDestroy).ngOnDestroy() call.

The one limitation today (we're working on it) is that if you inherit from a class that implements OnDestroy the lifecycle event is not called.

**Comment by matanlurey** [comment]
Blocked by https://github.com/dart-lang/angular2/issues/48

**Comment by matanlurey** [comment]
Just an update: Still blocked by #48 (same issue). _Early_ work is getting us towards this.

**Comment by matanlurey** [comment]
This is now planned for the 4.0 release.

**Comment by matanlurey** [comment]
Good news, this is now unblocked and we are testing a fix internally.

**Comment by matanlurey** [comment]
Blocked by https://github.com/dart-lang/angular2/issues/404.

**Comment by matanlurey** [comment]
Closing since it technically has landed. It will be available in `4.0.0.`

**Comment by erikhugintech** [comment]
This seems not to work if there's an extra class in between the component implementing life cycle classes and the class actually instantiated.

```
class Base implements OnInit
class Intermediate extends Base
class TheComponent extends Intermediate
```

ngOnInit call is not generated for TheComponent.

```
class Base implements OnInit
class Intermediate extends Base implements OnInit
class TheComponent extends Intermediate
```

Now ngOnInit call is generated for TheComponent.
